docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,237 @@
+   one of two buffers -- _valid_ and _invalid_.
+2. For transactions that DO NOT violate the nonce validation, they are included
+   in the _valid_ buffer.
+3. For transactions that DO violate the nonce validation, they are included in
+   the _invalid_ buffer.
+4. Continue this process until the desired number of valid transactions are
+   reaped or until the mempool is empty.
+5. Provide Tendermint the list of all transactions from the _valid_ buffer.
+6. Re-insert all transactions, from both buffers, back into app-side mempool.
+   This is to ensure we do not discard transactions from the app-side mempool in
+	 case `ProcessProposal` fails or in case that the proposal, while passing
+	 `ProcessProposal` is not the one decided for that height, i.e. the height took
+	 more than one round.
+
+```go
+type PriorityMempool[T MempoolTx] struct {
+	queue *PriorityQueue[MempoolTx]
+
+	// senders will contain a mapping from tx sender account addresses to all
+	// sequence numbers (nonces) or txs that they have in the app-side mempool.
+	senders map[string][]int64
+
+	// ...
+}
+```
+
+> The `PriorityMempool[T MempoolTx]` implementation will support Options such as
+> limiting the mempool size by a fixed number of bytes.
+
+Previous discussions<sup>1</sup> have come to the agreement that Tendermint will
+perform a request to the application, via `RequestPrepareProposal`, with a certain
+amount of transactions reaped from Tendermint's local mempool. The exact amount
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
+
+When Tendermint reaps transactions from the local mempool and sends them to the
+application via `RequestPrepareProposal`, the application will have to evaluate
+the transactions. Specifically, it will need to inform Tendermint if it should
+reject and or include each transaction. Note, the application can even _replace_
+transactions entirely with other transactions.
+
+When evaluating transactions from `RequestPrepareProposal`, the application will
+ignore _all_ transactions sent to it in the request and instead reap up to
+`RequestPrepareProposal.max_tx_bytes` from it's own mempool. There is no need to
+execute the transactions for validity as they have already passed CheckTx.
+
+### `ProcessProposal`
+
+The `ProcessProposal` ABCI method is relatively straightforward. It is responsible
+for ensuring validity of the proposed block containing transactions that were
+selected from the `PrepareProposal` step. In order to check validity of the proposed
+block, we must iterate over the list of transactions and execute them using the
+same mode/execution strategy as `CheckTx`. In other words, we execute each transaction
+using the AnteHandler only -- no messages are executed. However, we cannot just
+execute `CheckTx` again, because `BaseApp` already has a modified `checkState` at
```

Comment:
@alexanderbez Some food for thought here. Rather than utilizing the `AnteHandler` chain, would you be willing to explore a different API/approach for `ProcessProposal`? I have some concerns with automatically running all transactions through the `AnteHandlers` in `ProcessProposal`.

First, it's not clear to me that Cosmos SDK should be calling `CheckTx` and running every transaction in the proposal through the `AnteHandler` automatically. It adds a reasonable performance overhead given there is a global lock on the Tendermint side, and most of the time it is not necessary. For example, I don't believe the current implementation of Tendermint calls `CheckTx` for each transaction in the proposal either, and this seems to be fine?

Second, if you do move forward with this change, I think there could be potential unintended liveness implications for chains that do not opt-in to custom `ProcessProposal` behavior. My understanding is that today if an "invalid" transaction were to be proposed (say, with an incorrect sequence number or something like that) the `DeliverTx` response would return an error, but the transaction would still be included in the block. However, with your proposed changed this same transaction could lead to a rejected proposal which would increment the consensus round and delay the block. This is different liveness behavior than exists today and may not be desirable for all chains.

Instead, I view `ProcessProposal` as an API for Cosmos users who wish to explicitly add their own special bespoke logic to determine whether a proposal should be rejected/approved.

My initial assumption of how this would be included in Cosmos SDK was as a top-level method on the module ala `StartBlock`, and `EndBlock`, but something like the following:

```go
func (AppModule) BeginBlock(_ sdk.Context, _ abci.RequestBeginBlock) {}

func (AppModule) EndBlock(_ sdk.Context, _ abci.RequestEndBlock) []abci.ValidatorUpdate {}

func (AppModule) ProcessProposal(_ sdk.Context, _ abci.RequestProcessProposal) error {}
```

This would allow each module to decide how it wants to perform validation, and whether it deems it necessary to reject the proposal, which would potentially impact liveness of the chain.

I'm curious to get your thoughts on this approach.

Reply:
> First, it's not clear to me that Cosmos SDK should be calling CheckTx and running every transaction in the proposal through the AnteHandler automatically.

It won't be calling `CheckTx`, it's only calling the AnteHandler chain. Recall, `ProcessProposal` is solely accountable for ensuring the proposed block's validity.

> It adds a reasonable performance overhead given there is a global lock on the Tendermint side, and most of the time it is not necessary.

There is little to zero performance overhead. The entire point of the `AnteHandler` chain is to be lightweight, i.e. no message execution. Thats why `CheckTx` even exists.

> For example, I don't believe the current implementation of Tendermint calls CheckTx for each transaction in the proposal either, and this seems to be fine?

It does -- `DeliverTx` is a superset of `CheckTx`.

> Second, if you do move forward with this change, I think there could be potential unintended liveness implications for applications that do not have any requirements for custom ProcessProposal behavior.

Currently, as the ADR stands, `ProcessProposal` is not really customizable, where the `AnteHandler` chain is really the only customized part.

> However, now with this new proposed AnteHandler ProcessProposal approach, this same transaction would instead lead to a rejected proposal (because it would fail CheckTx in ProcessProposal) which would increment the consensus round and delay the block. This is different liveness behavior than exists today and may not be desirable for all applications.

Yes, that's correct. But I don't see this as an issue as we want to only propose correct blocks.

> Instead, I would view ProcessProposal as an API for Cosmos applications who wish to explicitly add their own special bespoke logic to determine whether a proposal should be rejected/approved.

Yes, the more I think about it, I think this is the better approach actually.

> `func (AppModule) ProcessProposal(_ sdk.Context, _ abci.RequestProcessProposal) error {}`

Yes, I think this is great. I will adjust the ADR.

Reply:
Updated. Please lmk what you think and we can resolve this convo.

Reply:
> It does -- DeliverTx is a superset of CheckTx.

Today `v0.34` of Tendermint calls `DeliverTx` only after the block is committed, but not while processing the proposal. Today the only validation done during proposal processing in Tendermint is the following (according to the [docs](https://docs.tendermint.com/master/spec/consensus/signing.html#proposals)):

```
p.Type == 0x20
p.Height > 0
p.Round >= 0
p.POLRound >= -1
p.BlockID.IsComplete()
```

> There is little to zero performance overhead. The entire point of the AnteHandler chain is to be lightweight, i.e. no message execution. Thats why CheckTx even exists.

That's fair. From our perspective, we are aiming for very high throughput, so we have been measuring blocking application performance by fractions of a millisecond. If we went the `AnteHandlers` route for `ProcessProposal`, we'd be running the transaction through all `AnteHandlers` 3 times instead of 2. There are occasional state (disk) reads occurring in the `auth` `AnteHandlers` for checking the sequence number and existence of the account in the `AccountsKeeper`. This is all I meant by "performance overhead". In practice these calls tend to be cached anyways, though so I agree with your "little to no" performance overhead assessment for most applications.

> Please lmk what you think and we can resolve this convo.

✅ LGTM. I also this think this solution will be de-risk things for you from a backwards compatibility standpoint as nothing should change for existing applications unless they explicitly opt in to `ProcessProposal` behavior in their applications by implementing this method in their modules.

Reply:
Great, I love the collaboration! Thanks for all the super valuable input :)